src/psyclone/psyGen.py

```diff
@@ -4705,6 +4968,300 @@ def _number_handler(self, node, parent):
+        '''
+        if new_scope not in Symbol.valid_scope_types:
+            raise ValueError("Symbol scope attribute can only be one of {0}"
+                             " but got '{1}'."
+                             "".format(str(Symbol.valid_scope_types),
+                                       str(new_scope)))
+
+        self._scope = new_scope
+
+    def __str__(self):
+        return (self.name + "<" + self.datatype + ", " + str(self.shape) +
+                ", " + self.scope + ">")
+
+
+class SymbolTable(object):
+    '''
+    Encapsulates the symbol table and provides methods to declare new symbols
+    and look up existing symbols. It is implemented as a single scope
+    symbol table (nested scopes not supported).
+    '''
+    def __init__(self):
+        # Dict of Symbol objects with the symbol names as keys.
+        self._symbols = {}
+        # Ordered list of the arguments.
+        self._argument_list = []
+
+    def declare(self, name, datatype, shape=[], scope='local',
+                is_input=False, is_output=False):
+        '''
+        Declare a new symbol in the symbol table.
+
+        :param str name: Name of the symbol.
+        :param str datatype: Datatype of the symbol.
+        :param list shape: Shape of the symbol in row-major order (leftmost \
+                       index is contiguous in memory). Each entry represents \
+                       an array dimension. If not None then it holds the \
+                       extent of that dimension. If it is an empy list it \
+                       represents an scalar.
+        :param str scope: It is 'local' if the symbol just exists inside the \
+                          kernel scope or 'global_*' if the data survives \
+                          outside of the kernel scope. Note that \
+                          global-scoped symbols also have postfixed \
+                          information about the sharing mechanism, at the \
+                          moment just 'global_argument' is available for \
+                          variables passed in/out of the kernel by argument.
+        :param bool is_input: Whether it is an already existent data that is \
+                              passed into the kernel upon entry.
+        :param bool is_output: Whether it is passed outside the kernel upon \
+                               exit.
+        :raises KeyError: The provided name can not be used as key in the
+                          table.
+        '''
+        if name in self._symbols:
+            raise KeyError("Symbol table already contains a symbol with"
+                           " name '{0}'.".format(name))
+
+        self._symbols[name] = Symbol(name, datatype, shape, scope, is_input,
+                                     is_output)
+
+    def specify_argument_list(self, argument_name_list):
+        '''
+        Keep track of the arguments order and provide the scope, is_input and
```

Comment:
"track of the order of the arguments and..."

Reply:
Done